compiler/rustc_middle/src/query/mod.rs

```diff
@@ -115,6 +118,12 @@ rustc_queries! {
         desc { "triggering a delayed bug for testing incremental" }
     }
 
+    query expand_legacy_bang(key: (LocalExpnId, LocalExpnId)) -> Result<(&'tcx TokenStream, usize), CanRetry> {
```

Comment:
these IDs are very easily going to change. Is there no better identifier? I guess the query is already `eval_always`, but it doesn't seem quite right. Intuitively I would have assumed some stable identifier for a macro and a `TokenStream` as input, which is that not possible?

Reply:
Do you mean using the macro's `DefId` and input `TokenStream` as identifier? I also think this is good, but `TokenStream` may contain `NonTerminal`, which is not hashable during the macro expansion stage. Unless we implement `Hash` for all ast stuffs.

Reply:
By the way, why are `LocalExpnId`s easier to change than `DefId`s?

Reply:
The `StableHash` impl on `DefId` doesn't actually hash the raw indices inside the `DefId`, but instead hashes the `DefPathHash` which is derived from the `DefPath`. The latter only changes when you rename items or their parent modules or when you have multiple items with the same name nested inside a single item. I believe the `LocalExpnId` contains that it is the Nth expansion in the current compilation session, which depends on both expansion order and all code that is expanded before the current macro expansion.

Reply:
> I also think this is good, but `TokenStream` may contain `NonTerminal`, which is not hashable during the macro expansion stage

I was unable to find `NonTerminal`. Also `TokenTree` implements `HashStable`, so it should be possible to stable-hash `TokenStream`, too, as it's just a wrapper around `Lrc<Vec<TokenTree>>`.

Reply:
FYI, nonterminals are currently being removed in https://github.com/rust-lang/rust/pull/124141.
After that token streams will no longer contain pieces of AST.

Reply:
There's analogue of `DefPathHash` for expansion ids called `ExpnHash`.
It was introduced in https://github.com/rust-lang/rust/pull/87044 and https://github.com/rust-lang/rust/pull/86676 (you may be interested in reading those threads).

This hash is stable in the sense that it is suitable for working with multiple compilations, but it is probably less resistant to small changes in code than `DefPathHash` (https://github.com/rust-lang/rust/pull/86676#issuecomment-877489460).
Although it may be more resistant now than back in 2021 because spans going into the incremental system are item-relative now.

Getting performance improvements from incremental expansion reuse *may* require using some other more change-resistant hashing scheme.

Reply:
>Only starting from HIR, right?

Not anymore:
https://github.com/rust-lang/rust/blob/master/compiler/rustc_expand/src/expand.rs#L597-L604

Reply:
I'd rather use LocalExpnId here, or at least leave a fixme. 

Making LocalExpnId as stable as DefId (or almost) is a long term goal, and the signature should use it when we reach that state.